mne/artifact_detection.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Nov 19 09:40:37 2019
+
+@author: an512
+"""
+
+from scipy.stats import zscore
+import mne
+import numpy as np
+
+from mne.annotations import Annotations
+from scipy.ndimage.measurements import label
+from mne.io.ctf.trans import _quaternion_align
+from mne.chpi import _apply_quat
+from itertools import compress
+
+
+def detect_bad_channels(raw, zscore_v=4, method='both', t1=30, t2=220,
+                        neigh_max_distance=.035):
+    """ detect bad channels based on z-score amplitude deviation or/ and
+        decreased local correlation with other channels"""
+
+    # set recording length
+    Fs = raw.info['sfreq']
+    t2x = min(raw.last_samp / Fs, t2)
+    t1x = max(0, t1 + t2x - t2)  # Start earlier if recording is shorter
+
+    # Get data
+    raw_copy = raw.copy().crop(t1x, t2x).load_data()
+    raw_copy = raw_copy.pick_types(meg=True, ref_meg=False)\
+        .filter(1, 45).resample(150, npad='auto')
+    data_chans = raw_copy.get_data()
+
+    # Get channel distances matrix
+    chns_locs = np.asarray([x['loc'][:3] for x in
+                            raw_copy.info['chs']])
+    chns_dist = np.linalg.norm(chns_locs - chns_locs[:, None],
+                               axis=-1)
+    chns_dist[chns_dist > neigh_max_distance] = 0
+
+    # Get avg channel uncorrelation between neighbours
+    chns_corr = np.abs(np.corrcoef(data_chans))
+    weig = np.array(chns_dist, dtype=bool)
+    chn_nei_corr = np.average(chns_corr, axis=1, weights=weig)
+    chn_nei_uncorr_z = zscore(1 - chn_nei_corr)  # l ower corr higer Z
+
+    # Get channel magnitudes
+    max_Pow = np.sqrt(np.sum(data_chans ** 2, axis=1))
+    max_Z = zscore(max_Pow)
+
+    if method == 'corr':  # Based on local uncorrelation
```

Comment:
this is not validated is it? I'm not sure we should put in MNE methods that are not validated. On the other hand, if you provide all the knobs and transparently document what the method is with an example, maybe it's not so bad.

Reply:
Well, it is just a zscored local correlation, a bad channel will be measuring uncorrelated activity from its peers